_posts/2020-01-24-inception-pipeline-cdk.md

```diff
@@ -0,0 +1,395 @@
+      adminPermissions: true,
+      // Hard coded as cdk only synthesises the template. The deployment is separate.
+      templatePath: cdkBuildOutput.atPath(`${NPRFES}.template.json`),
+      stackName: `${NPRFES}`,
+      region: props.nonProd.env.region,
+      runOrder: 1,
+    }),
+    new codepipeline_actions.CodeBuildAction({
+      actionName: "BuildReactApp",
+      project: buildDeployReactNonProd,
+      input: sourceOutput,
+      runOrder: 2,
+    }),
+  ],
+},
+```
+
+But where is that template coming from? Glad you asked. It's part of the synthesised template which
+we generate the build cdk stage of the pipeline. The artifacts array needs to be updated:
+
+```typescript
+artifacts: {
+  "base-directory": "cdj-build-pipeline/dist",
+  files: [
+    `${this.stackName}.template.json`,
+    `${CRSS}.template.json`,
+    `${NPRFES}.template.json`,
+  ],
+},
+```
+
+The eagle-eyed reader among us will notice some strange string substitution going on here, let me
+explain. When the CDK synthesises it will generate stacks for everything that is required. This
+includes sub-stacks (our front-end stack, for example), as well as supporting stacks that are needed
+in destination accounts in order for the deployments to work.
+
+```typescript
+const NPRFES = "NonProdReactFrontEndStack";
+const CASS = "cross-account-support-stack-<account number>";
+```
+
+The CASS variable I've invented above is the name I've given to the template that is generated by a
+synthesis of the CDK application we're working with.
+{:refdef: style="text-align: center;"}
+<img src="/img/cdk_synth_out.png">
+{: refdef}
+
+In order to facilitate the CloudFormation step which would deploy infrastructure in our destination
+account, we'll need to give the destination account permissions and roles for the CloudFormation
+step to assume in order to be able to do the things it needs to to deploy. This is where our
+intermediate step comes in for the BuildAndAdministerPipeline stage.
+
+So the revised stage now looks like this:
+
+```typescript
+stageName: "BuildAndAdministerPipeline",
+actions: [
+  new codepipeline_actions.CodeBuildAction({
+    actionName: "CDK_Build",
+    project: cdkBuild,
+    input: sourceOutput,
+    outputs: [cdkBuildOutput],
+    runOrder: 1,
+  }),
+  new codepipeline_actions.CloudFormationCreateUpdateStackAction({
+    actionName: "ProvisionCrossAccountSupportNonProd",
+    account: props.nonProd.env.account,
+    templatePath: cdkBuildOutput.atPath(`${CASS}.template.json`),
+    stackName: "CrossAccountSupport",
+    adminPermissions: true,
+    runOrder: 2,
+  }),
+  new codepipeline_actions.CloudFormationCreateUpdateStackAction({
+    actionName: "AdministerPipeline",
+    templatePath: cdkBuildOutput.atPath(`${this.stackName}.template.json`),
+    stackName: this.stackName,
+    adminPermissions: true,
+    runOrder: 3,
+  }),
+],
+```
+
+Before this stage can run the first time, it needs to be deployed in the destination account once,
```

Comment:
Could potentially do this with a StackSet?

Reply:
Not really, because the destination account still needs perms to do stuff, which then needs to be self-managing. I am advising to deploy the stack which manages the permissions the first time in the destination account so subsequent updates can be self-managing